lib/clients/GcpClient.py

```diff
@@ -0,0 +1,553 @@
+                    '{} SUCCESS: snapshot-id={}'.format(
+                        log_prefix, snapshot_id))
+                return True
+            else:
+                message = '{} ERROR: snapshot-id={}, snapshot still exists'.format(
+                    log_prefix, snapshot_id)
+                self.logger.error(message)
+                raise Exception(message)
+        except Exception as error:
+            message = '{} ERROR: snapshot-id={}\n{}'.format(
+                log_prefix, snapshot_id, error)
+            self.logger.error(message)
+            raise Exception(message)
+
+    def _create_volume(self, size, snapshot_id=None):
+        log_prefix = '[VOLUME] [CREATE]'
+        volume = None
+
+        try:
+            disk_name = self.generate_name_by_prefix(self.disk_prefix)
+            disk_body = {
+                'name': disk_name,
+                'sizeGb': size,
+                'labels': self.tags
+                # default typ is pd-standard
+            }
+            if snapshot_id:
+                disk_body['sourceSnapshot'] = 'global/snapshots/{}'.format(
+                    snapshot_id)
+
+            self.logger.info('Creating disk with details {}'.format(disk_body))
+
+            disk_creation_operation = self.compute_client.disks().insert(
+                project=self.project_id, zone=self.availability_zone, body=disk_body).execute()
+
+            self._wait('Waiting for volume {} to get ready...'.format(disk_name),
+                       (lambda operation_id, zonal_operation: self.get_operation_status(
+                           operation_id, zonal_operation) == 'DONE'),
+                       None,
+                       disk_creation_operation['name'], True)
+
+            volume = self.get_volume(disk_name)
+            self._add_volume(volume.id)
+
+            self.logger.info(
+                '{} SUCCESS: volume-id={} with tags = {} '.format(log_prefix, volume.id, self.tags))
+        except Exception as error:
+            message = '{} ERROR: volume-id={}, size={}\n{}'.format(
+                log_prefix, disk_name, size, error)
+            self.logger.error(message)
+            self.delete_volume(disk_name)
+            volume = None
+            raise Exception(message)
+
+        return volume
+
+    def _delete_volume(self, volume_id):
+        log_prefix = '[VOLUME] [DELETE]'
+
+        try:
+            disk_deletion_operation = self.compute_client.disks().delete(
+                project=self.project_id, zone=self.availability_zone, disk=volume_id).execute()
+
+            self._wait('Waiting for disk {} to be deleted...'.format(volume_id),
+                       (lambda operation_id, zonal_operation: self.get_operation_status(
+                           operation_id, zonal_operation) == 'DONE'),
+                       None,
+                       disk_deletion_operation['name'], True)
+
+            volume_exists = self.volume_exists(volume_id)
```

Comment:
If delete operation returns success. do you still need to cross check for existence of disk?

Reply:
Ideally, not needed. But added it as an extra precaution as we might have dangling volumes if not deleted properly.